### PR TITLE
Update Resolution Failures to report to Problems API

### PR DESCRIPTION
**Pull request by tresat**
Replaces https://github.com/gradle/gradle/pull/29520 to incorporate the changes from https://github.com/gradle/gradle/pull/29581.

Adds separate display name for each problem id.

**https://github.com/gradle/gradle/pull/29581 should be merged first**

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@donat - Do you think it's worth adding to some of the existing test expectations in `ResolutionFailureIntegrationTest` that the proper failures were reported?  

Currently, it seems like there's no way for them _not to have been reported_, but maybe there is an easy way to check this?

**Comment by donat** [comment]
Yes, absolutely. You can easily add problems API checks by calling `enableProblemsApiCheck()` during test setup. If you run the test with the check enabled the test should fail and the output will contain [hints](https://github.com/gradle/gradle/blob/7def017d5e7d892565e8690d756d330a3f77b1c6/testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy#L836) on the missing assertions.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test and merge

File: subprojects/core/src/main/java/org/gradle/api/internal/catalog/problems/ResolutionFailureProblemId.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    NO_COMPATIBLE_VARIANTS("No variants exist that would match the request"),
+    NO_VARIANTS_WITH_MATCHING_CAPABILITIES("No variants exist with capabilities that would match the request"),
+
+    // Stage 3 failures
+    AMBIGUOUS_ARTIFACT_TRANSFORM("Multiple artifacts transforms exist that would satisfy the request"),
+    NO_COMPATIBLE_ARTIFACT("No artifacts exist that would match the request"),
+    AMBIGUOUS_ARTIFACTS("Multiple artifacts exist that would match the request"),
+    UNKNOWN_ARTIFACT_SELECTION_FAILURE("Unknown artifact selection failure"),
+
+    // Stage 4 failures
```
Comment (jvandort) [inline subprojects/core/src/main/java/org/gradle/api/internal/catalog/problems/ResolutionFailureProblemId.java:44]: We should avoid the stage terminology

**Pull request merged**